runtime/jvmti/jvmtiHelpers.c

```diff
@@ -95,56 +95,110 @@ static UDATA watchedClassEqual (void *lhsEntry, void *rhsEntry, void *userData);
+	BOOLEAN isThreadAlive = FALSE;
+#if JAVA_SPEC_VERSION >= 19
+	BOOLEAN isVirtualThread = FALSE;
+#endif /* JAVA_SPEC_VERSION >= 19 */
 
-	if (thread == NULL) {
+	if (NULL == thread) {
 		if (allowNull) {
 			*vmThreadPtr = currentThread;
 			return JVMTI_ERROR_NONE;
 		}
 		return JVMTI_ERROR_INVALID_THREAD;
 	} else {
-		threadObject = *((j9object_t*) thread);
+		threadObject = J9_JNI_UNWRAP_REFERENCE(thread);
 		if (currentThread->threadObject == threadObject) {
 			*vmThreadPtr = currentThread;
 			return JVMTI_ERROR_NONE;
 		}
 	}
 
-	/* Make sure the vmThread stays alive while it is being used */
-
+	/* Make sure the vmThread stays alive while it is being used. */
 	omrthread_monitor_enter(vm->vmThreadListMutex);
-	if (!J9VMJAVALANGTHREAD_STARTED(currentThread, threadObject) || ((targetThread = J9VMJAVALANGTHREAD_THREADREF(currentThread, threadObject)) == NULL)) {
+#if JAVA_SPEC_VERSION >= 19
+	isVirtualThread = IS_VIRTUAL_THREAD(currentThread, threadObject);
+	if (isVirtualThread) {
+		if  (NULL != isVirtual) {
+			*isVirtual = TRUE;
+		}
+		/* Make sure the virtual thread stays alive while it is being used. */
+		omrthread_monitor_enter(vm->liveVirtualThreadListMutex);
+		jint vthreadState = J9VMJAVALANGVIRTUALTHREAD_STATE(currentThread, threadObject);
+		j9object_t carrierThread = (j9object_t)J9VMJAVALANGVIRTUALTHREAD_CARRIERTHREAD(currentThread, threadObject);
+		if (NULL != carrierThread) {
+			targetThread = J9VMJAVALANGTHREAD_THREADREF(currentThread, carrierThread);
+		}
+		isThreadAlive = (vthreadState != JVMTI_VTHREAD_STATE_NEW) && (vthreadState != JVMTI_VTHREAD_STATE_TERMINATED);
+	} else
+#endif /* JAVA_SPEC_VERSION >= 19 */
+	{
+		targetThread = J9VMJAVALANGTHREAD_THREADREF(currentThread, threadObject);
+		isThreadAlive = J9VMJAVALANGTHREAD_STARTED(currentThread, threadObject) && (NULL != targetThread);
+	}
+
+	if (!isThreadAlive) {
 		if (mustBeAlive) {
+#if JAVA_SPEC_VERSION >= 19
+			if (isVirtualThread) {
+				omrthread_monitor_exit(vm->liveVirtualThreadListMutex);
+			}
+#endif /* JAVA_SPEC_VERSION >= 19 */
 			omrthread_monitor_exit(vm->vmThreadListMutex);
 			return JVMTI_ERROR_THREAD_NOT_ALIVE;
 		}
 	}
 
 	*vmThreadPtr = targetThread;
-	if (targetThread != NULL) {
-		++(targetThread->inspectorCount);
+	if (NULL != targetThread) {
+		targetThread->inspectorCount += 1;
+	}
+#if JAVA_SPEC_VERSION >= 19
+	if (isVirtualThread) {
+		U_64 vthreadInspectorCount = J9OBJECT_U64_LOAD(currentThread, threadObject, vm->virtualThreadInspectorCountOffset);
+		J9OBJECT_U64_STORE(currentThread, threadObject, vm->virtualThreadInspectorCountOffset, (vthreadInspectorCount + 1));
+		omrthread_monitor_exit(vm->liveVirtualThreadListMutex);
 	}
+#endif /* JAVA_SPEC_VERSION >= 19 */
 	omrthread_monitor_exit(vm->vmThreadListMutex);
 	return JVMTI_ERROR_NONE;
 }
 
 
 
 void
-releaseVMThread(J9VMThread * currentThread, J9VMThread * targetThread)
+releaseVMThread(J9VMThread *currentThread, J9VMThread *targetThread, jthread thread, BOOLEAN isVirtual)
```

Comment:
does isVritual need to be a parameter?

Reply:
There are few cases where the caller doesnt care. If you pass null to `getVMThread` then what do you pass into `releaseVMThread`

Reply:
> does isVritual need to be a parameter?

No. But, `IS_VIRTUAL_THREAD` will be called more than twice. In some cases, `isVirtual` will also be used in the core body of the JVMTI function (e.g. [GetThreadInfo](https://github.com/eclipse-openj9/openj9/issues/15183#issuecomment-1187580033)). This approach shares the result of `IS_VIRTUAL_THREAD` after evaluating it once in `getVMThread`.

> There are few cases where the caller doesnt care. If you pass null to getVMThread then what do you pass into releaseVMThread

Those cases will only handle platform threads from `J9VMThread->carrierThreadObject`. Virtual threads are not expected. `getVMThread` will be given `NULL`, and `releaseVMThread` will be given `FALSE`.

Reply:
In `jvmtiGetOSThreadID` for example. The `isVirtual` variable is created but is never actaully needed. Is it possible for `releaseVMThread` to fetch the data from `thread` ?

Reply:
Such scenarios need to be skipped for yielded virtual threads `(targetThread == NULL)`. `jvmtiInterruptThread` also falls in this category. These functions

Reply:
by `thread`, I mean `jthread thread` which is just an object.

Reply:
what are the conditions where the threadObject can be null in `releaseVMThread`

Reply:
I guess what Im saying is:
if `thread` exists, then you can also just unwrap it and determine if its a virtual thread and do the necesarry parts for `vthreadInspectorCount`

If it doesnt exist, then presumably it wasn't a virtual thread or at the very least you wont need to do anything for `vthreadInspectorCount` since the counters are on the object itself

Reply:
if a user sends a NULL `thread` (as input) and `getVMThread` allows a `null`. in such a case, `isVirtual` will stay `FALSE`, and none of the new code will be invoked.

Reply:
> I dont understand this comment

The below code will segfault when a virtual thread is yielded. Because `targetThread` will be null for a yielded thread. We need to protect it with `(targetThread == NULL)`.
```
		rc = getVMThread(currentThread, thread, &targetThread, TRUE, TRUE, &isVirtual); <- Will pin the j.l.VirtualThread
		if ((rc == JVMTI_ERROR_NONE) && (NULL == targetThread)) { <- Protect for null targetThread
			rv_threadid = (jlong) omrthread_get_osId(targetThread->osThread);
			releaseVMThread(currentThread, targetThread, thread, isVirtual); <- Won't be executed for null targetThread so the j.l.VirtualThread won't be unpinned
		}
```

In other JVMTI functions, we will need to stack-allocate a `J9VMThread` if a virtual thread is yielded.

Reply:
> if a user sends a NULL thread (as input) and getVMThread allows a null. in such a case, isVirtual will stay FALSE, and none of the new code will be invoked.

So doesnt this mean you dont need the `isVirtual` parameter? Because if `thread` is NULL then `isVirtual` must be false. And if `thread` is not null, you can always determine if it is virtual.

Reply:
And Im only talking about `releaseVMThread`.

Reply:
`BOOLEAN isVirtual` allows us to share the result of `IS_VIRTUAL_THREAD` after evaluating it once in `getVMThread`. `IS_VIRTUAL_THREAD == isSameOrSuperClassOf` will be invoked multiple times if `BOOLEAN isVirtual` is removed.

Reply:
Yes, I understand. My issue is that the `isVirtual` is relevant in `getVMThread` and `releaseVMThread`, but not always in the caller. It makes it confusing to define variables that have no meaning in the context that they are defined in. 

I think it would be cleared for `jvmtiGetOSThreadID` to be

```
	rc = getVMThread(currentThread, thread, &targetThread, TRUE, TRUE);
		rc = getVMThread(currentThread, thread, &targetThread, TRUE, TRUE, NULL);
		if (rc == JVMTI_ERROR_NONE) {
			rv_threadid = (jlong) omrthread_get_osId(targetThread->osThread);
			releaseVMThread(currentThread, targetThread);
			releaseVMThread(currentThread, targetThread, thread);
```

Reply:
Updated.